modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/TransactionStorage.scala

```diff
@@ -96,51 +114,100 @@ class TransactionStorage[F[_]: Async: KryoSerializer](
-            Consecutive.take[F](waiting.toNonEmptyList.toList, lastTx)
+            val lastTxState = lastAccepted.getOrElse(address, Majority(TransactionReference.empty))
+            Consecutive
+              .take[F](waiting.toList, lastTxState.ref)
+              .map(consecutiveTxs => pullForAddress(lastTxState, consecutiveTxs))
           case None =>
             List.empty[Hashed[Transaction]].pure[F]
         }
       }.map(_.flatten)
     } yield txs.size
 
-  def pull(): F[Option[NonEmptyList[Hashed[Transaction]]]] =
+  def pull(count: NonNegLong): F[Option[NonEmptyList[Hashed[Transaction]]]] =
     for {
       lastAccepted <- lastAccepted.toMap
-      addresses <- waitingTransactions.keys
+      addresses <- waitingTransactions.keys.map(_.sorted)
       allPulled <- addresses.traverse { address =>
         val pulledM = for {
           (maybeWaiting, setter) <- waitingTransactions(address).access
 
-          maybePulled <- maybeWaiting.traverse { waiting =>
-            val lastTx = lastAccepted.getOrElse(address, TransactionReference.empty)
+          pulled <- maybeWaiting.traverse { waiting =>
+            val lastTxState = lastAccepted.getOrElse(address, Majority(TransactionReference.empty))
             Consecutive
-              .take[F](waiting.toList, lastTx)
-              .map { pulled =>
-                (SortedSet.from(waiting.toList.diff(pulled)).toNes, pulled)
-              }
+              .take[F](waiting.toList, lastTxState.ref)
+              .map(consecutiveTxs => pullForAddress(lastTxState, consecutiveTxs))
+              .map(pulled => (SortedSet.from(waiting.toList.diff(pulled)).toNes, pulled))
               .flatMap {
                 case (maybeNotPulled, pulled) =>
                   val maybeStillWaiting = maybeNotPulled
-                    .flatMap(_.filter(_.ordinal > lastTx.ordinal).toNes)
+                    .flatMap(_.filter(_.ordinal > lastTxState.ref.ordinal).toNes)
 
                   setter(maybeStillWaiting)
                     .ifM(
-                      logger.debug(s"Pulled ${pulled.size} transaction(s) for consensus").as(pulled),
+                      NonEmptyList.fromList(pulled).pure[F],
                       logger.debug("Concurrent update occurred while trying to pull transactions") >>
-                        List.empty[Hashed[Transaction]].pure[F]
+                        none[NonEmptyList[Hashed[Transaction]]].pure[F]
                     )
               }
-          }
-          pulled = maybePulled.getOrElse(List.empty)
+          }.map(_.flatten)
         } yield pulled
 
         pulledM.handleErrorWith {
           logger.warn(_)(s"Error while pulling transactions for address=${address.show} from waiting pool.") >>
-            List.empty.pure[F]
+            none[NonEmptyList[Hashed[Transaction]]].pure[F]
         }
       }.map(_.flatten)
-    } yield NonEmptyList.fromList(allPulled)
+
+      selected = takeFirstNHighestFeeTxs(allPulled, count)
```

Comment:
should we sort by fee? that way there is an option that some transactions will be never pulled 🤔

Reply:
that's the whole point I guess of having transactions with fee, that they have the precedence

Reply:
We could of course add a functionality to always take pulled feeless transactions but a consequence will be either a possibility to block fee transactions (not making the rate limit per block) or we will loose the control over limit of transactions per block (as we will need to pull over the limit). This in general touches upon very high network traffic and I'm not sure that there is a solution that would solve all the problems.